### PR TITLE
[Keyboard] Add keyboard Reviung39

### PR DESCRIPTION
**Pull request by gtips**
Add Reviung39 keyboard.

## Description
The REVIUNG39 is 39-key (37.5%) column staggered keyboard.

Soldering is divided into through-hole mounting version and surface mounting (SMD) version so that the soldering difficulty can be selected.

- REVIUNG39 is regular version (SMD)
- REVIUNG39S is simple version (through-hole)

Both have the same key layout, design and size.
However, the number of RGB LEDs to be mounted is different.
That's why there are two default keymaps.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/reviung39/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#define _BASE 0
+#define _LOWER 1
+#define _RAISE 2
+#define _ADJUST 16
```
Comment (fauxpark) [inline keyboards/reviung39/keymaps/default/keymap.c]: Non-consecutive layer numbers will generate extra empty layers in between, unnecessarily increasing flash usage. The better way to define layers is with an enum, eliminating the need to number them manually:

```c
enum layer_names {
    _BASE,
    _LOWER,
    _RAISE,
    _ADJUST
};
```

File: keyboards/reviung39/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#define LOWER  MO(_LOWER)
+#define RAISE  MO(_RAISE)
+#define ADJUST MO(_ADJUST)
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_BASE] = REVIUNG39_LAYOUT( \
```
Comment (fauxpark) [inline keyboards/reviung39/keymaps/default/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

File: keyboards/reviung39/reviung39.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/* This a shortcut to help you visually see your layout.
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+#define REVIUNG39_LAYOUT( \
```
Comment (fauxpark) [inline keyboards/reviung39/reviung39.h]: QMK naming convention for this layout macro would be `LAYOUT_reviung39`

File: keyboards/reviung39/rules.mk [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# MCU name
+#MCU = at90usb1286
```
Comment (fauxpark) [inline keyboards/reviung39/rules.mk]:
```suggestion
```

File: keyboards/reviung39/rules.mk [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+MCU = atmega32u4
+
+# Processor frequency.
```
Comment (fauxpark) [inline keyboards/reviung39/rules.mk]: Lines 5 to 39 can be removed, QMK provides these same values as defaults.

File: keyboards/reviung39/rules.mk [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   atmega32a    bootloadHID
+#BOOTLOADER = atmel-dfu
```
Comment (fauxpark) [inline keyboards/reviung39/rules.mk]:
```suggestion
```

File: keyboards/reviung39/keymaps/default/keymap.c (hunk 1 above)
Comment (gtips) [reply keyboards/reviung39/keymaps/default/keymap.c]: thanks

File: keyboards/reviung39/keymaps/default/keymap.c (hunk 2 above)
Comment (gtips) [reply keyboards/reviung39/keymaps/default/keymap.c]: thanks

File: keyboards/reviung39/reviung39.h (hunk 3 above)
Comment (gtips) [reply keyboards/reviung39/reviung39.h]: thanks

File: keyboards/reviung39/rules.mk (hunk 4 above)
Comment (gtips) [reply keyboards/reviung39/rules.mk]: thanks

File: keyboards/reviung39/rules.mk (hunk 5 above)
Comment (gtips) [reply keyboards/reviung39/rules.mk]: thanks

File: keyboards/reviung39/rules.mk (hunk 6 above)
Comment (gtips) [reply keyboards/reviung39/rules.mk]: thanks

File: keyboards/reviung39/keymaps/default/keymap.c [hunk 7]
```diff
@@ -15,42 +15,44 @@
...
   ),
   
-  [_ADJUST] = REVIUNG39_LAYOUT(\
-    RGB_VAI,   RGB_SAI, RGB_HUI,  RGB_MOD,  XXXXXXX,   RGB_TOG,            KC_EJCT,  XXXXXXX,  XXXXXXX,  KC_BRMU,  KC__VOLUP, KC_VOLU,  \
-    RGB_VAD,   RGB_SAD, RGB_HUD,  RGB_RMOD, XXXXXXX,   XXXXXXX,            XXXXXXX,  XXXXXXX,  XXXXXXX,  KC_BRMD,  KC__VOLDOWN,KC_VOLD, \
-    XXXXXXX,   XXXXXXX, XXXXXXX,  XXXXXXX,  XXXXXXX,   XXXXXXX,            RESET,    XXXXXXX,  XXXXXXX,  XXXXXXX,  KC__MUTE,  KC_MUTE,  \
-                                                       _______,  XXXXXXX,  _______   \
+  [_ADJUST] = LAYOUT_reviung39(
+    RGB_VAI,   RGB_SAI, RGB_HUI,  RGB_MOD,  XXXXXXX,   RGB_TOG,            KC_EJCT,  XXXXXXX,  XXXXXXX,  KC_BRMU,  KC__VOLUP, KC_VOLU,
+    RGB_VAD,   RGB_SAD, RGB_HUD,  RGB_RMOD, XXXXXXX,   XXXXXXX,            XXXXXXX,  XXXXXXX,  XXXXXXX,  KC_BRMD,  KC__VOLDOWN,KC_VOLD,
```
Comment (fauxpark) [inline keyboards/reviung39/keymaps/default/keymap.c]: No need to use `KC_BRMU/D` and `KC__VOLUP/DOWN`, these are macOS specific, whereas `KC_BRIU/D` and `KC_VOLU/D` will work on all OSes. Same with `KC_MUTE`.

Comment (gtips) [reply keyboards/reviung39/keymaps/default/keymap.c]: thanks

Comment (gtips) [reply keyboards/reviung39/keymaps/default/keymap.c]: Rethinking that volume control is not necessary, I removed it from the default keymap.

**Comment by drashna** [comment]
Thanks!

**Pull request merged**